Convert policy unit tests to use incremental path

fristonio opened the pull request:
Make policy unit tests to use incremental path with policy resolution validation on Identity add and delete.

fristonio committed:
test/policy: extend policy unit tests for incremental path

Signed-off-by: Deepesh Pathak <deepeshpathak09@gmail.com>

joestringer commented on pkg/policy/l4.go:
Curious if this is only used for testing, could we put it in `l4_test.go` instead?

squeed commented on pkg/policy/l4_filter_test.go:
```diff
@@ -117,6 +139,98 @@ func (td *testData) addIdentity(id *identity.Identity) {
+			return true
+		}
+
+		require.Equal(t, l4.Port, l4B.Port)
```
Why not use `L4Filter.Equals()`?

squeed replied:
If this is because the PerSelectorPolicies don't match, is there a way you can find to keep them matching?

fristonio replied:
Its because in the PerSelectorPolicyMap the key is the CachedSelector pointer. When we use `L4Filter.Equal` we  match that the pointers are equal. Earlier we were using the same cached selector ptr in expected and actual objects which we would insert to `SelectorCache` during init.
Now, the expected and actual objects don't point to the same cached selector as we also have incremental id add/delete, so we have this custom impl to make sure cached selectors are equal in the map.

squeed replied:
Ah. Should we add an `Equals()` to the CachedSelector interface?

fristonio replied:
That wouldn't necessarily help us in this case no? For incremental update we still need a custom implementation for checking if the selectors selects the same set of identities based on what is upserted in the selector cache.
`L4Filter.Equals` is a test only function anyways. We can remove the original `Equal` and rely on this custom test equal if we are concerned about the drift in behavior. Diff is also a little bit cleaner with this check implemented in tests.

squeed commented on pkg/policy/l4_filter_test.go:
```diff
@@ -117,6 +139,98 @@ func (td *testData) addIdentity(id *identity.Identity) {
+
+			require.True(t, found, "Failed to find expected cached selector in PerSelectorPolicy: %s", k.String())
+		}
+
```
I'd like to add an additional code path that converts the expected L4PolicyMap to a MapState with a full SelectorCache, then converts the actual with no peer identities.  Then, add them incrementally and verify the MapStates are the same.

derailed commented on pkg/policy/l4_filter_test.go:
```diff
@@ -45,6 +47,8 @@ type testData struct {
 	sc   *SelectorCache
 	repo *Repository
 
+	idSet *set.Set[identity.NumericIdentity]
```
A set (map) is already a reference. Why is this needed vs `set.Set[...]?

derailed commented on pkg/policy/l4_filter_test.go:
```diff
@@ -70,31 +74,45 @@ func newTestData(logger *slog.Logger) *testData {
 	td := &testData{
 		sc:                testNewSelectorCache(logger, nil),
 		repo:              NewPolicyRepository(logger, nil, &fakeCertificateManager{}, envoypolicy.NewEnvoyL7RulesTranslator(logger, certificatemanager.NewMockSecretManagerInline()), nil, testpolicy.NewPolicyMetricsNoop()),
+		idSet:             &set.Set[identity.NumericIdentity]{},
```
nit: Prefer set.New[...]()

fristonio committed:
policy: fix l4 selector handling for incremental id updates

This commit fixes the deviation in processing l4 selectors for a filter
during incremental id updates. With this change, if the selector being
processed for the filter is a subset of the wildcard selector, we skip
processing the map state change. This is how map state evaluation is
handled in endpoint policy resolution path.
This commit ensures that the final policy map state is consistent in
both incremental identity update and selector policy distillation path.

Signed-off-by: Deepesh Pathak <deepeshpathak09@gmail.com>

squeed commented on pkg/policy/l4.go:
```diff
@@ -628,6 +628,47 @@ func (c *ChangeState) Size() int {
 }
 
+// generateWildcardMapStateEntry creates map state entry for wildcard selector in the filter.
+func (l4 *L4Filter) generateWildcardMapStateEntry(logger *slog.Logger, p *EndpointPolicy, port uint16) *mapStateEntry {
```
nit: @jrajahalme observed that making `mapStateEntry` a pointer significantly impacts benchmarks, since it requires a heap allocation. That's one reason we use `entry.Invalid`

In fact, do we really need this function at all?

fristonio replied:
Oh, thats interesting. TIL!
The function is just to reduce boilerplate, nothing functional. I have fixed the changes to avoid the extra allocation. Added a bench test for the same. I would prefer to keep the function to avoid duplicating the code, lmk what you think.

#### `main`
<img width="1274" height="156" alt="image" src="https://github.com/user-attachments/assets/25c5e45d-f6a0-444f-94c6-7161211eb4d3" />

#### Fix
<img width="1353" height="150" alt="image" src="https://github.com/user-attachments/assets/f23a5c3c-7e3f-48fa-8cb9-74f141b7fbcf" />
